Fixes terminal completion command for odo

cdrage opened the pull request:
<!--
Thank you for opening a PR! Here are some things you need to know before submitting:

1. Please read our developer guideline: https://github.com/redhat-developer/odo/wiki/Dev:-odo-Dev-Guidelines
2. Label this PR accordingly with the '/kind' line
3. Ensure you have written and ran the appropriate tests: https://github.com/redhat-developer/odo/wiki/Dev:-Writing-and-running-tests
4. Read how we approve and LGTM each PR: https://github.com/redhat-developer/odo/wiki/Pull-Requests:-Review-guideline

Documentation:

If you are pushing a change to documentation, please read: https://github.com/redhat-developer/odo/wiki/Documentation:-Contributing
-->

**What type of PR is this:**

<!--
Add one of the following kinds:
/kind bug
/kind feature
/kind tests
/kind documentation

Feel free to use other [labels](https://github.com/redhat-developer/odo/labels) as needed. However one of the above labels must be present or the PR will not be reviewed. This instruction is for reviewers as well.
-->
/kind cleanup

**What does this PR do / why we need it:**

This PR:
* Changes the command `odo utils terminal` to `odo completion` similar
  to kubectl.
* Re-enables / fixes the terminal completion command support for bash,
  zsh, fish and powershelll

**Which issue(s) this PR fixes:**
<!--
Specifying the issue will automatically close it when this PR is merged
-->

Fixes #5536

**How to test changes / Special notes to the reviewer:**

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

cdrage commented:
If people can please test please do!

The instructions are at `odo completion --help` for bash, zsh, powershell and fish.

valaparthvi commented:
@cdrage, do you think you can add a documentation for this command?

valaparthvi commented on pkg/odo/cli/completion/completion.go:
```diff
@@ -0,0 +1,93 @@
+	# POWERSHELL
+
+	## Load into your current shell environment
+	%[1]s completion powershell | Out-String | Invoke-Expression
+
+	## Load persistently
+	%[1]s completion powershell >> $PROFILE
```
```suggestion
	%[1]s powershell | Out-String | Invoke-Expression

	## Load persistently
	%[1]s powershell >> $PROFILE
```

rm3l commented:
@cdrage I reproduced what I was mentioning during the Standup call. Unlike the other additional flags, `-o` does not seem to be listed in the completion list. Don't know if that is the expected behavior with this flag.
I am using zsh, FYI.

Example with `odo analyze`:
```
❯ odo analyze -<TAB>
--kubeconfig      -- Paths to a kubeconfig. Only required if out-of-cluster.                                                                                                                                                                                                   
--v           -v  -- number for the log level verbosity. Level varies from 0 to 9 (default 0).                                                                                                                                                                                 
--vmodule         -- comma-separated list of pattern=N settings for file-filtered logging
```

And here is the help output of `odo analyze`, which mentions the `-o` option.:
```
❯ odo analyze --help 
Detect devfile to use based on files present in current directory

Usage:
  odo analyze [flags]

Flags:
  -h, --help   Help for analyze

Additional Flags:
      --kubeconfig string    Paths to a kubeconfig. Only required if out-of-cluster.
  -o, --o string             Specify output format, supported format: json
  -v, --v Level              Number for the log level verbosity. Level varies from 0 to 9 (default 0).
      --vmodule moduleSpec   Comma-separated list of pattern=N settings for file-filtered logging
```

Same behavior with `odo describe component` or `odo describe binding`.

cdrage commented:
> @cdrage I reproduced what I was mentioning during the Standup call. Unlike the other additional flags, `-o` does not seem to be listed in the completion list. Don't know if that is the expected behavior with this flag. I am using zsh, FYI.
> 
> Example with `odo analyze`:
> 
> ```
> ❯ odo analyze -<TAB>
> --kubeconfig      -- Paths to a kubeconfig. Only required if out-of-cluster.                                                                                                                                                                                                   
> --v           -v  -- number for the log level verbosity. Level varies from 0 to 9 (default 0).                                                                                                                                                                                 
> --vmodule         -- comma-separated list of pattern=N settings for file-filtered logging
> ```
> 
> And here is the help output of `odo analyze`, which mentions the `-o` option.:
> 
> ```
> ❯ odo analyze --help 
> Detect devfile to use based on files present in current directory
> 
> Usage:
>   odo analyze [flags]
> 
> Flags:
>   -h, --help   Help for analyze
> 
> Additional Flags:
>       --kubeconfig string    Paths to a kubeconfig. Only required if out-of-cluster.
>   -o, --o string             Specify output format, supported format: json
>   -v, --v Level              Number for the log level verbosity. Level varies from 0 to 9 (default 0).
>       --vmodule moduleSpec   Comma-separated list of pattern=N settings for file-filtered logging
> ```
> 
> Same behavior with `odo describe component` or `odo describe binding`.

Ah! That's a bug, I'll find it. Thank you. Odd that it's only `-o` doing it.

cdrage commented:
/retest

feloy commented on pkg/odo/cli/cli.go:
```diff
@@ -152,10 +152,6 @@ func odoRootCmd(name, fullName string) *cobra.Command {
 
-	// We will mark the command as hidden and then re-enable if the command
-	// supports json output
-	_ = pflag.CommandLine.MarkHidden("o")
```
By removing this, the option now appears in all commands, event if they do not support it.

cdrage commented:
@feloy @rm3l 

Fixed!

Basically... I've unhidden the -o flag, but I've modified the help usage page to hide it if it's not supported.

Essentially, we cannot "unhide" and "hide" flags programatically within Cobra in order to generate the correct bash completion.

Unfortunately, with the way that completion is handled within cobra, it'll still show for each command.

```sh
go/src/odo  fix-utils ✗                                                                                                                                                                                                                                                                                                                                         3d21h ⚑
▶ odo deploy --o json
 ✗  Machine readable output is not yet implemented for this command
```

Ex. For when trying to run it.

I suggest that we still merge this in, and create an issue upstream (Cobra) as well as an issue within our repo with a WONTFIX until it's fixed upstream.

feloy commented:
You will need to fix these "validate" errors:

```
pkg/odo/cli/completion/completion.go:72:33: Error return value of `(*github.com/spf13/cobra.Command).GenBashCompletion` is not checked (errcheck)
				cmd.Root().GenBashCompletion(os.Stdout)
				                            ^
pkg/odo/cli/completion/completion.go:79:14: Error return value of `out.Write` is not checked (errcheck)
				out.Write([]byte(zsh))
				         ^
pkg/odo/cli/completion/completion.go:80:32: Error return value of `(*github.com/spf13/cobra.Command).GenZshCompletion` is not checked (errcheck)
				cmd.Root().GenZshCompletion(out)
				                           ^
pkg/odo/cli/completion/completion.go:82:33: Error return value of `(*github.com/spf13/cobra.Command).GenFishCompletion` is not checked (errcheck)
				cmd.Root().GenFishCompletion(os.Stdout, true)
				                            ^
pkg/odo/cli/completion/completion.go:84:47: Error return value of `(*github.com/spf13/cobra.Command).GenPowerShellCompletionWithDesc` is not checked (errcheck)
				cmd.Root().GenPowerShellCompletionWithDesc(os.Stdout)
```

cdrage commented:
Ready for review 👍 @rm3l @valaparthvi

rm3l reviewed (approved):
Tested with both ZSH and Bash, and completion works great 🎉

One side note: we could have added this to the documentation (command reference and/or somewhere in the installation guide). For now, this seems hidden as a feature.

/approve

feloy commented:
/override ci/prow/unit
/override ci/prow/v4.10-integration-e2e

feloy commented:
/override ci/prow/unit
/override ci/prow/v4.10-integration-e2e

cdrage commented:
> 

You're right, I'll add a doc for that!
